### DISCUSSION TITLE
HttpClient vs YARP Direct Forwarding

### DISCUSSION THREAD

**Discussion by jjkcharles** [dotnet/yarp#2220] [Q&A] [answered]
I vaguely remember reading somewhere about suggesting to use YARP Direct Forwarding for performance reasons in cases where an API needs to call another API without much business logic. I couldn't locate the source, nor do I recollect where I came across it.

Recently, our team stumbled upon a very similar scenario and I was wondering may be should this be one of the options to optimize the API. But, since YARP internally uses HTTPClient anyways, I was little doubtful as to the performance gain it can offer.

So, I went on to test it out and see how they both compare. [This is the repo I created with a minimal sample](https://github.com/jjkcharles/Yarp-HttpClient-Performance-Analysis) - as can be seen in readme.md, the test results produced have very similar results for both cases with non-YARP method having better throughput & response time as the load increases.

Wanted to understand,
1. If what I "assume" I have read somewhere is really a valid statement or not?
2. If valid, why isn't my sample providing better results when using YARP? What changes are needed to make it better?

Sample Repo URL: https://github.com/jjkcharles/Yarp-HttpClient-Performance-Analysis

**Comment by Tratcher** [comment]
If direct forwarding is adequate for your needs then yes, it can save a few layers of abstraction/middleware. However, this isn't going to show up in the perf benchmarks for your average application, your scale would have to be huge for it to be relevant.

**Reply by Tratcher** [reply, marked as answer]
Wait, sorry, I completely misread that 🥱😂. Direct forwarding will only be faster than using HttpClient directly if our implementation is more optimized than yours. We do lots of tricks like streaming, zero-byte reads, buffer pooling, etc. that your average application wouldn't attempt. You also need to hit a pretty high scale before it's relevant.

**Reply by jjkcharles** [reply]
Good to know, thanks. Appreciate you looking into this!
